src/core/public/fatal_errors/fatal_errors_service.tsx

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { render } from 'react-dom';
+import * as Rx from 'rxjs';
+import { first, tap } from 'rxjs/operators';
+
+import { InjectedMetadataService } from '../injected_metadata';
+import { FatalErrorsScreen } from './fatal_errors_screen';
+import { ErrorInfo, getErrorInfo } from './get_error_info';
+
+export interface FatalErrorsParams {
+  rootDomElement: HTMLElement;
+  injectedMetadata: InjectedMetadataService;
+  stopCoreSystem: () => void;
+}
+
+export class FatalErrorsService {
+  private errorInfo$ = new Rx.ReplaySubject<ErrorInfo>(Infinity);
+
+  constructor(private params: FatalErrorsParams) {
+    this.errorInfo$.pipe(first(), tap(this.onFirstError)).subscribe({
+      error: error => {
+        // tslint:disable-next-line no-console
+        console.error('Uncaught error in fatal error screen internals', error);
+      },
+    });
+  }
+
+  public add = (error: Error | string, source?: string) => {
+    const errorInfo = getErrorInfo(error, source);
+
+    this.errorInfo$.next(errorInfo);
+
+    if (error instanceof Error) {
+      // make stack traces clickable by putting whole error in the console
+      // tslint:disable-next-line no-console
+      console.error(error);
+    }
+
+    throw error;
+  };
+
+  public start() {
+    return {
+      add: this.add,
+      get$: () => {
+        return this.errorInfo$.asObservable();
+      },
+    };
+  }
+
+  private onFirstError = () => {
```

Comment:
I agree with using `tap(() => this.onFirstError())`, `tap(this.onFirstError)` seems to hide what is happening just a little bit, and doesn't provide any real benefits. 

For `add: this.add` I prefer to leave it this way so that TypeScript can know the correct types without having to redeclare them, and so when I "go to definition" on that method elsewhere it takes me to the real definition, not just some wrapper defined in `start()`, but I suppose I could be persuaded differently.